### ISSUE TITLE
[torch.compile] returns NaN for `tensor.mul(big_number).softmax()`

### ISSUE THREAD

**Issue by payphon** [pytorch/pytorch#100775]
### 🐛 Describe the bug

`torch.compile` returns NaN for `tensor.mul(big_number).softmax()`

```py
import torch

torch.manual_seed(420)

class Model(torch.nn.Module):

    def __init__(self, input_size, hidden_size):
        super(Model, self).__init__()
        self.query = torch.nn.Linear(input_size, hidden_size)
        self.key = torch.nn.Linear(input_size, hidden_size)
        self.scale_factor = torch.nn.Parameter(torch.tensor([-1.7197e+14]))

    def forward(self, x):
        query = self.query(x)
        key = self.key(x)
        scores = torch.matmul(query, key.transpose(-2, -1))
        scores_scaled = scores.mul(self.scale_factor)
        t = scores_scaled.softmax(dim=-1)
        return t

input_size = 32
hidden_size = 1

x = torch.randn(10, input_size)
func = Model(input_size, hidden_size)

jit_func = torch.compile(func)

res1 = func(x) # without jit
res2 = jit_func(x)
print(res1)
# tensor([[0.], ...)
print(res2)
# tensor([[nan], ...)
```

### Versions

```
Collecting environment information...
PyTorch version: 2.1.0.dev20230503+cu118
Is debug build: False
CUDA used to build PyTorch: 11.8
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.1 LTS (x86_64)
GCC version: (Ubuntu 11.3.0-1ubuntu1~22.04) 11.3.0
Clang version: 14.0.0-1ubuntu1
CMake version: Could not collect
Libc version: glibc-2.35

Python version: 3.9.16 (main, Mar  8 2023, 14:00:05)  [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.19.5-051905-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 11.5.119
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 3060
Nvidia driver version: 510.108.03
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Address sizes:                   46 bits physical, 48 bits virtual
Byte Order:                      Little Endian
CPU(s):                          24
On-line CPU(s) list:             0-23
Vendor ID:                       GenuineIntel
Model name:                      12th Gen Intel(R) Core(TM) i9-12900K
CPU family:                      6
Model:                           151
Thread(s) per core:              2
Core(s) per socket:              16
Socket(s):                       1
Stepping:                        2
CPU max MHz:                     6700.0000
CPU min MHz:                     800.0000
BogoMIPS:                        6374.40
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault cat_l2 invpcid_single cdp_l2 ssbd ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid rdt_a rdseed adx smap clflushopt clwb intel_pt sha_ni xsaveopt xsavec xgetbv1 xsaves split_lock_detect avx_vnni dtherm ida arat pln pts hwp hwp_notify hwp_act_window hwp_epp hwp_pkg_req hfi umip pku ospke waitpkg gfni vaes vpclmulqdq tme rdpid movdiri movdir64b fsrm md_clear serialize pconfig arch_lbr ibt flush_l1d arch_capabilities
Virtualization:                  VT-x
L1d cache:                       640 KiB (16 instances)
L1i cache:                       768 KiB (16 instances)
L2 cache:                        14 MiB (10 instances)
L3 cache:                        30 MiB (1 instance)
NUMA node(s):                    1
NUMA node0 CPU(s):               0-23
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Mmio stale data:   Not affected
Vulnerability Retbleed:          Not affected
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Enhanced IBRS, IBPB conditional, RSB filling, PBRSB-eIBRS SW sequence
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Not affected

Versions of relevant libraries:
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.24.2
[pip3] pytorch-triton==2.1.0+7d1a95b046
[pip3] torch==2.1.0.dev20230503+cu118
[pip3] torchaudio==2.1.0.dev20230503+cu118
[pip3] torchvision==0.16.0.dev20230503+cu118
[conda] numpy                     1.24.2                   pypi_0    pypi
[conda] pytorch-triton            2.1.0+7d1a95b046          pypi_0    pypi
[conda] torch                     2.1.0.dev20230503+cu118          pypi_0    pypi
[conda] torchaudio                2.1.0.dev20230503+cu118          pypi_0    pypi
[conda] torchvision               0.16.0.dev20230503+cu118          pypi_0    pypi
```

cc @ezyang @msaroufim @wconstab @bdhirsh @anijain2305 @zou3519 @soumith @ngimel

**Comment by XiaobingSuper** [comment]
This is FMA gives a precision gap, see the generated code:
```
 {
        #pragma GCC ivdep
        for(long i0=static_cast<long>(0L); i0<static_cast<long>(10L); i0+=static_cast<long>(1L))
        {
            {
                float tmp_acc0 = 0;
                for(long i1=static_cast<long>(0L); i1<static_cast<long>(10L); i1+=static_cast<long>(1L))
                {
                    auto tmp0 = in_ptr0[static_cast<long>(i1 + (10L*i0))];
                    auto tmp1 = in_ptr1[static_cast<long>(0L)];
                    auto tmp3 = out_ptr0[static_cast<long>(i0)];
                    auto tmp2 = decltype(tmp0)(tmp0 * tmp1);
                    auto tmp4 = tmp2 - tmp3;
                    auto tmp5 = std::exp(tmp4);
                    out_ptr1[static_cast<long>(i1 + (10L*i0))] = tmp5;
                    tmp_acc0 += tmp5;
                }
                out_ptr2[static_cast<long>(i0)] = tmp_acc0;
            }
        }
    }
''')
```

When gcc using ```-O3 -march=native```, the compiler will have some optimization according to machine instruction subsets, you can test it by running https://godbolt.org/z/b4K8o9qqh by setting a different gcc optimization flag, for example, if you are using ```-O0```, there will not have ```nan``` number.

**Comment by XiaobingSuper** [comment]
cc @jgong5

**Comment by jgong5** [comment]
We might have to disable automatic fp contraction by compiler (`-ffp-contract=off`) since fma usually results in higher accuracy than non-fused operations, therefore the result doesn't match and causes wrong result in this case. I'm just not sure the perf impact of disable this. Needs benchmark results to know..

**Comment by leslie-fang-intel** [comment]
Need double confirm the performance data.

**Comment by Valentine233** [comment]
For validation of disabling fp-contraction flag, FP32's result is ok, but BF16's result is with doubt. Several models have regressions for eager mode, which is not expected. The regression can't be locally reproduced on SPR machine.

Would relaunch the BF16 validation using AWS instance with bare metal instead of virtual machine.

**Comment by Valentine233** [comment]
According to the latest BF16 validation result with bare metal, there is no abnormal model regressions for eager mode.

**Comment by swolchok** [comment]
> https://godbolt.org/z/b4K8o9qqh 

This looks like a pathological case for FMA -- that example relies on being able to predict the intermediate a * b result exactly, which of course you can't do with FMA. Perhaps we could modify our generated code for softmax so that it doesn't rely on this property, rather than disabling FMA globally? It seems bad that our generated code for linear and mm doesn't use FMAs.

**Comment by jgong5** [comment]
> > https://godbolt.org/z/b4K8o9qqh
> 
> This looks like a pathological case for FMA -- that example relies on being able to predict the intermediate a * b result exactly, which of course you can't do with FMA. Perhaps we could modify our generated code for softmax so that it doesn't rely on this property, rather than disabling FMA globally? It seems bad that our generated code for linear and mm doesn't use FMAs.

Not sure if we can do specific things for `softmax` since it has already been decomposed when being codegen. On the other hand, perhaps we can explicitly generate fma (something like ops.fma) for operations that need it?
